### ISSUE TITLE
[Ingester] Add ability to completely flush on shutdown

### ISSUE THREAD

**Issue by joe-elliott** [grafana/tempo#699]
**Is your feature request related to a problem? Please describe.**
When an ingester shuts down it is expected it will come back up, replay its wal and continue using the same disk. However, in some ephemeral deployments this is not true and it is desired for an ingester to always flush its entire contents to the backend on shutdown.

Add the ability for an ingester to always flush its entire disk to the backend.

Note: With this setting there may be gaps in trace availability. After the ingester shuts down but before the queriers get a chance to poll the new blocks those traces would not be available. This should be noted on the configuration option, but does not need to be solved.

**Comment by dgzlopes** [comment]
Started researching this :eyes:

Just to double-check that I understood the ticket properly.

We want to add some config option, that when enabled, makes ingesters flush all in-memory traces and the WAL to the long-term backend on shutdown. A behavior that is somewhat similar to the current `/shutdown` [endpoint](https://grafana.com/docs/tempo/latest/api_docs/#shutdown). Right?

**Comment by joe-elliott** [comment]
Yes, but I don't think that the /shutdown endpoint is quite right. This loop for instance:
```
for !i.flushQueues.IsEmpty() {
	time.Sleep(100 * time.Millisecond)
}
```
Probably has a chance of exiting even when all work is not done. When the complete operation is being handled and before the flush operation is queued the queues will be empty. Unsure. This issue is more difficult then it seems on its surface and requires deep diving the ingester flushing mechanisms. Please reach out for help!

**Comment by dgzlopes** [comment]
Gotcha! I'll try to dig deeper into those mechanisms. Sounds like a good opportunity to learn more about ingesters :D

**Comment by mghildiy** [comment]
If no-one is looking into it, I may try to look how tempo ingesters work.

**Comment by joe-elliott** [comment]
Sure, thanks for taking a look. This function here actually does the job of flushing and waiting for all queues to drain:

https://github.com/grafana/tempo/blob/main/modules/ingester/flush.go#L91

So you'd need to work this into the shutdown logic here:

https://github.com/grafana/tempo/blob/main/modules/ingester/ingester.go

**Comment by mghildiy** [comment]
A question here:
What decides ingester node to be ephemeral? Is there any configurarion? And does it mean that it needs to be checked? Or we  just flush to backend irrespective of anything?

**Comment by joe-elliott** [comment]
This should not be default behavior as it will increase the time it takes for the ingesters to shut down. Most operators will prefer to shut down and then come back up using the same disk and not use this feature.

Something like?
```
ingester:
  flush_all_on_shutdown:
```

**Comment by mghildiy** [comment]
So when user deploys tempo on a node, this config would need to be configured accordingly. By default, its value is kept false.

**Comment by joe-elliott** [comment]
Right, this would only be used in exceptional cases. For instance, I believe we had a community member running Tempo on Fargate and wanted functionality like this.

**Comment by mghildiy** [comment]
What user action OR cause can actually lead to ingester shutdown?

**Comment by joe-elliott** [comment]
Generally the application will receive an OS level signal called "SIGTERM" which indicates it needs to start shutting down. Currently the ingester uses this signal to stop receiving traces and flush all data to disk. With this change it would also attempt to flush all data to object storage as well.

In Kubernetes every time it attempts to stop a pod it will send SIGTERM to all containers in the pod. For instance if you were to do a rollout of your ingesters: `kubectl rollout restart statefulset/ingester` they would receive SIGTERM one at a time as k8s rolled them out.

**Comment by mghildiy** [comment]
It may take some time as I try to understand lots of things invovled here...ingester, traces, instance, backends etc. Things are quite connected, it seems so far.

**Comment by mghildiy** [comment]
'Add the ability for an ingester to always flush its entire disk to the backend.'

I guess what it means is to write all of the WAL content to the configured backend(GCS, Azure,AWS, local).

**Comment by mghildiy** [comment]
I have a question.

During sweepAllInstances execution, sweepInstance is invoked on every instance. This in turn invokes instance.CutBlockIfReady, which in the process calls WALBlock.Flush(). Flush is documented as:

```
// Flush any unbuffered data to disk.  Must be safe for concurrent use with read operations.
Flush() error
```

So this function flushes data into disk. 
What does disk here represent? Is it WAL or parquet-file on local hard drive? Or backend maybe?

Also, [here](https://grafana.com/docs/tempo/latest/configuration/), I see 'flush_check_period', which talks about 

`how often to sweep all tenants and move traces from live -> wal -> completed blocks.`

I understand live here is in-memory, WAL is write-ahead-log. But I am not sure what exactly completed blocks is?

Any inputs here would help.

**Comment by mghildiy** [comment]
In ShutdownHandler, I see:

```
// move all data into flushQueue
i.sweepAllInstances(true)

for !i.flushQueues.IsEmpty() {
     time.Sleep(100 * time.Millisecond)
}
```

sweepAllInstances basically converts headblock for instances into completing blocks(ie blocks which can be flushed into disk), and each such block marked is enqueued into flushQueues to be processed further.

i.flushQueues.IsEmpty waits for all entries in flushQueues to be processed. But I see flushQueues to be processed only inside flushLoop function, which is not part of above code flow(but rather part of service starting flow). It makes me wonder what actually causes entries in flushQueues to be processed in ShutdownHandler flow.

**Comment by joe-elliott** [comment]
> What does disk here represent? Is it WAL or parquet-file on local hard drive? Or backend maybe?

"Disk" there means flushing to the local block storage. The wal is kept on local disk and that method is used to cause anything in memory to be written to disk.

> I understand live here is in-memory, WAL is write-ahead-log. But I am not sure what exactly completed blocks is?

Completed blocks are kept on disk until the `completed_block_timeout` has expired. These have been written to backend storage but they remain on disk so that the ingesters can answer queries about these traces until the queriers are guaranteed to pick up their existence.

> i.flushQueues.IsEmpty waits for all entries in flushQueues to be processed. But I see flushQueues to be processed only inside flushLoop function, which is not part of above code flow(but rather part of service starting flow). It makes me wonder what actually causes entries in flushQueues to be processed in ShutdownHandler flow.

This flushloop is running in a separate goroutine and is responsible for servicing the queues. https://github.com/grafana/tempo/blob/main/modules/ingester/flush.go#L199

**Comment by mghildiy** [comment]
```
Completed blocks are kept on disk until the completed_block_timeout has expired. These have been written to backend storage but they remain on disk...
```

So it means completed blocks are not required to be written to backend storage as its already done. Only completing blocks and headblock are to be send to backend.

**Comment by mghildiy** [comment]
Apart from disk, live traces too need to be pushed to backend I guess. Basically every trace available in the instance that is not yet pushed.

**Comment by joe-elliott** [comment]
> Apart from disk, live traces too need to be pushed to backend I guess. Basically every trace available in the instance that is not yet pushed.

Yup. I think that code in the `ShutdownHandler` does that. Do you think differently?

**Comment by mghildiy** [comment]
If ShutdownHandler could take care of even livetraces then we wont have any problem if I understand code correctly, because completing blocks stage comes only after live traces are cut into head block. 

ShutdownHandler invokes sweepAllInstances, which ensures that livetraces and completing blocks available so far are made available to be flushed into backend. But it uses the following check to 'try' to ensure that all flush operations created by  sweepAllInstances are considered:

```
for !i.flushQueues.IsEmpty() {
			time.Sleep(100 * time.Millisecond)
}
```

But as flushloop routines are running in background, I think there may be a window possible when this check fails and we still have flush operations in queue not processed. Put succintly, reliance on Sleep is not a good idea I think. Its very similar to Java where we dont rely on sleep for such cases, and it seems same in go concurrency. 

But I can be totally wrong, thanks to my limited knowledge of Golang and Golang concurrency model.

One thing that crossed my mind. Why not stop all concurrent processing for stop/shutdown scenarios, and then do everything sequentially to ensure that everything is taken care of. Just an idea, but it may be a very terrible suggestion.

**Comment by joe-elliott** [comment]
> But as flushloop routines are running in background, I think there may be a window possible when this check fails and we still have flush operations in queue not processed. Put succintly, reliance on Sleep is not a good idea I think. Its very similar to Java where we dont rely on sleep for such cases, and it seems same in go concurrency.

This is a good point. We manually inspect logs when using the ShutdownHandler so it "works" but agree with automatic shutdown we would want something more reliable. 

The main issue I see is that if we pull a complete job from the queue here:
https://github.com/grafana/tempo/blob/06d43bf4a96c44dd44bc40bf779a0d890e0354e6/modules/ingester/flush.go#L206

Then handle it  here:
https://github.com/grafana/tempo/blob/06d43bf4a96c44dd44bc40bf779a0d890e0354e6/modules/ingester/flush.go#L217

Then we enqueue a flush job for the now completed block here:
https://github.com/grafana/tempo/blob/06d43bf4a96c44dd44bc40bf779a0d890e0354e6/modules/ingester/flush.go#L288

If the `i.flushQueues.Empty()` check happens to hit in between pulling a complete job and pushing the flush job it will return true even if there is still work being done.

> One thing that crossed my mind. Why not stop all concurrent processing for stop/shutdown scenarios, and then do everything sequentially to ensure that everything is taken care of. Just an idea, but it may be a very terrible suggestion.

I'm not sure what this would look like off the top of my head, but I like the idea.

**Comment by mghildiy** [comment]
But we invoke sweelAllInstances with argument true(hence all processing happen immediately) before making i.flushQueues.isEmpty() check.
So I am not sure if check would have false positive.
Let me dig more in the code before I come to any decisive conclusion.

**Comment by mghildiy** [comment]
https://github.com/grafana/tempo/blob/06d43bf4a96c44dd44bc40bf779a0d890e0354e6/modules/ingester/flush.go#L207-L209

This would return from the flushLoop function, and hence terminate corresponding thread/go routine, isn't..while we want them to be alive so that tasks are handled whenver available.

**Comment by mghildiy** [comment]
activekeys is a Map to keep track of operations due for processing. Its key is detemined by (op kind + userid + blockid)

sweepAllInstances adds opKindComplete task to queue, and update the Map activekeys

flushLoop flow does:

- Takes out the task using dequeue operation. But dequeue operation doesnt modify activekeys, thus still holding correpsonding key
- Depending upon type of operation, flushes to backend when its type opKindFlush or adds new task of type opKindFlush to the queue when task is type opKindComplete(updating the activekeys in the process)
- If task, irrespective of the type, is successful, activeKeys is cleared of the corresponding key.


i.flushQueues.isEmpty check uses activeKeys to take the decision. activeKeys is cleared of a key only when its either successful or we dont get a retry signal, which seems to be good noeuigh logical conclusion. 
Hence I think i.flushQueues.isEmpty should always give correct picture of the state.

**Comment by mghildiy** [comment]
A WIP [PR](https://github.com/grafana/tempo/pull/2515).

**Comment by mghildiy** [comment]
[Updated PR](https://github.com/grafana/tempo/pull/2538)

**Comment by mghildiy** [comment]
For testing the new method, spying is needed. I am not sure if thats possible in go. Let me know if any other approach is possible, menawhile I try if spying can be done.
